pennylane_lightning/lightning_qubit.py

```diff
@@ -138,7 +116,10 @@ def apply(self, operations, rotations=None, **kwargs):
             self._pre_rotated_state = self._state
 
         if rotations:
-            self._state = self.apply_lightning(self._pre_rotated_state, rotations)
+            if any(isinstance(r, QubitUnitary) for r in rotations):
+                super().apply(operations=[], rotations=rotations)
+            else:
+                self._state = self.apply_lightning(np.copy(self._pre_rotated_state), rotations)
```

Comment:
This is getting me wondering: each time we call `self.apply_lightning`, we are doing conversion between Python and C++. If we combined `operations` and `rotations`, then we could get away with only a single conversion. In exchange, it would not be so simple to obtain the `_pre_rotated_state` as before.

Reply:
Agreed! Could we just return the pre-rotated state and the post-rotated state? Also I recall some discussion on the rotations API changing, but can't remember the decision?

Reply:
> Could we just return the pre-rotated state and the post-rotated state?

We should be able to, though it might require some thought to 1. still minimize the number of copies made of the state vector 2. keep the code readable.

A naive way would be to make the bound `apply` C++ function return the pre-rotated state *and* change the state vector in place to yield the post-rotated state. However, wondering if such a function would lead to much confusion because we'd basically return an intermediate result of the computation.

Also unsure at this moment on how much overhead we're actually incurring from calling on bound C++ functions twice.

> Also I recall some discussion on the rotations API changing, but can't remember the decision?

Not quite sure about this one. 🤔